File: sliding_window/3_longest_substring_without_repeating_characters.cpp
#include <string>
#include <iostream>
#include <unordered_set>
using namespace std;


/* Given a string s, find the length of the longest substring without repeating characters.
https://leetcode.com/problems/longest-substring-without-repeating-characters/
O(n), O(n) */

class Solution {
public:
    int lengthOfLongestSubstring(string& s) {
        int res = 0, l = 0;
        unordered_set<char> sSet;
        for (int r = 0; r < s.size(); r++) {
            while (sSet.find(s[r]) != sSet.end()) {
                sSet.erase(s[l]);
                l++;
            }
            sSet.insert(s[r]);
            res = max(res, r - l + 1);
        }
        return res;
    }
};



int main (int argc, char *argv[])
{
    string s = "pwwkew";
    cout << Solution().lengthOfLongestSubstring(s) << endl;
    return 0;
}

File: sliding_window/3_longest_substring_without_repeating_characters.py
"""
Given a string s, find the length of the longest substring without repeating characters.
https://leetcode.com/problems/longest-substring-without-repeating-characters/
O(n), O(n)
"""


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        charSet = set()
        l = 0
        res = 0
        for r in range(len(s)):
            while s[r] in charSet:
                charSet.remove(s[l])
                l += 1
            charSet.add(s[r])
            res = max(res, r - l + 1)
        return res


if __name__ == "__main__":
    # s = "abcabcbb"
    s = "pwwkew"
    print(Solution().lengthOfLongestSubstring(s))
